### DISCUSSION TITLE
Why does firejail block multiple invocations of librewolf?

### DISCUSSION THREAD

**Discussion by crocket** [netblue30/firejail#4538] [Q&A] [answered]
Because librewolf is a distribution of firefox, this issue applies to firefox as well.

My bookmark manager opens each of multiple chosen bookmarks with an invocation of `librewolf --new-tab URL`. Thus, it calls `librewolf --new-tab URL` multiple times simultaneously if I try to open multiple bookmarks in the bookmark manager.

This librewolf.local allows me to open one new tab at a time.

```
dbus-user.own org.mozilla.librewolf.*

# enable native notifications
dbus-user.talk org.freedesktop.Notifications

# Allow disabling screensaver
dbus-user.talk org.freedesktop.ScreenSaver

# screensharing under wayland
whitelist ${RUNUSER}/pipewire-0
whitelist /usr/share/pipewire/client.conf
dbus-user.talk org.freedesktop.portal.*
```

But, if I try to open several new tabs at a time, librewolf becomes unresponsive and quits with an error after a while. This is the error.

> LibreWolf is already running, but is not responding. To use LibreWolf, you must first close the existing LibreWolf process, restart your device, or use a different profile.

If I don't sandbox librewolf with firejail, there is no error when opening multiple new librewolf tabs at once.

Disabling dbus filter in a sandbox doesn't fix the issue. Is there anything else that I have to whitelist?

**Comment by kmk3** [comment]
@crocket [on Sep 13](https://github.com/netblue30/firejail/discussions/4538#discussion-3568082):

> Because librewolf is a distribution of firefox, this issue applies to firefox
> as well.
> 
> My bookmark manager opens each of multiple chosen bookmarks with an
> invocation of `librewolf --new-tab URL`. Thus, it calls `librewolf --new-tab
> URL` multiple times simultaneously if I try to open multiple bookmarks in the
> bookmark manager.

[...]

> But, if I try to open several new tabs at a time, librewolf becomes
> unresponsive and quits with an error after a while. This is the error.
> 
> > LibreWolf is already running, but is not responding. To use LibreWolf, you must first close the existing LibreWolf process, restart your device, or use a different profile.
> 
> If I don't sandbox librewolf with firejail, there is no error when opening
> multiple new librewolf tabs at once.
> 
> Disabling dbus filter in a sandbox doesn't fix the issue. Is there anything
> else that I have to whitelist?

Does the following help?

```
--join-or-start=name
      Join the sandbox identified by name or start a new one.  Same as
      "firejail --join=name" if sandbox with specified name exists, otherwise
      same as "firejail --name=name ..." Note that in contrary to other join
      options there is respective profile option.
```

Example:

```sh
firejail --join-or-start=librewolf123 librewolf --new-tab URL
```

Note: If you have multiple librewolf profiles, you can use a different sandbox
name for each one.

**Reply by crocket** [reply]
I tried to open 5 bookmarks with `--join-or-start=librewolf`, and I got

```
$ firejail --list
21989:user:librewolf:/usr/bin/firejail --join-or-start=librewolf --profile=librewolf /usr/bin/librewolf --new-tab http://www.gavo.t.u-tokyo.ac.jp/ojad/phrasing 
21991:user:librewolf-21991:/usr/bin/firejail --join-or-start=librewolf --profile=librewolf /usr/bin/librewolf --new-tab http://www.gavo.t.u-tokyo.ac.jp/ojad/search 
21993:user:librewolf-21993:/usr/bin/firejail --join-or-start=librewolf --profile=librewolf /usr/bin/librewolf --new-tab https://www.wadoku.de/ 
21995:user:librewolf-21995:/usr/bin/firejail --join-or-start=librewolf --profile=librewolf /usr/bin/librewolf --new-tab https://forvo.com/ 
21997:user:librewolf-21997:/usr/bin/firejail --join-or-start=librewolf --profile=librewolf /usr/bin/librewolf --new-tab https://jisho.org 
```

![2021-09-13 15:49:51](https://user-images.githubusercontent.com/748856/133037089-95160c0e-78f4-4672-9dd2-8c3e258314eb.png)

`--join-or-start` doesn't really make 4 other invocations join `librewolf` sandbox.

It takes roughly 100 milliseconds to launch a sandbox, and my bookmark manager is trying to open 5 bookmarks simultaneously. My bookmark manager doesn't wait for the first invocation to create a sandbox.

**Reply by rusty-snake** [reply]
@crocket you must have started your running instance of librewolf with `join-or-start`.

**Reply by crocket** [reply]
Perhaps, I should modify my bookmark manager so that it creates librewolf sandbox when there is none and joins it when there is one.

**Reply by rusty-snake** [reply]
OT: Page and extensions loading is racy. Meaning if you do `firefox/librewolf http://example.com` and there is no running instance, `http://example.com` might get loaded before AddOns like uBlock origin, uMatrix or NoScript are loaded.

**Reply by crocket** [reply]
Without dbus filtering,

* If I try to open multiple bookmarks in the absence of any librewolf instance, only the first bookmark is opened, and the other bookmarks are blocked.
* If I try to open multiple bookmarks in the presence of one librewolf instance, I can open as many bookmarks as I want.

With dbus filtering and `dbus-user.own org.mozilla.librewolf.*`, I can only open one bookmark at a time with or without an existing librewolf instance.

**Reply by rusty-snake** [reply]
Can you try to `--ignore=private-tmp`.

**Reply by crocket** [reply]
With `--ignore=private-tmp` and `dbus-user.own org.mozilla.librewolf.*`, opening multiple bookmarks simultaneously in the absence of an existing librewolf instance is possible, but a few librewolf instances are blocked from exitting after giving remote command and launch error messages after a while. If I launch multiple bookmarks simultaneously in the presence of one librewolf instance, there is no issue. `dbus-user.own org.mozilla.librewolf.*` seems to lead to a deadlock when I try to open multiple instances of librewolf simultaneously in the absence of an existing librewolf instance.

With `--ignore=private-tmp` and without dbus filtering, opening multiple bokmarks simultaneously is possible, and all subsequent librewolf instances quit without blocking after giving remote command to the first instance.

**Reply by kmk3** [reply]
@crocket [on Sep 13](https://github.com/netblue30/firejail/discussions/4538#discussioncomment-1314708):

> With `--ignore=private-tmp` and without dbus filtering, opening multiple
> bokmarks simultaneously is possible, and all subsequent librewolf instances
> quit without blocking after giving remote command to the first instance.

Do you use dbus for anything on librewolf?

Does `dbus-user none` make it any better/worse?

**Reply by crocket** [reply]
An existing librewolf instance gets new tab URLs on `org.mozilla.librewolf.PROFILE-ID`. Thus, DBus is necessary for using librewolf with bookmark manager.

**Reply by rusty-snake** [reply]
Firefox uses D-Bus (`org.mozilla.<PROGRAM>.Base64(<PROFILE-NAME>)`) as IPC backend if it is a native Wayland client. If it is a native X11 client, it uses some X11 magic based IPC.

https://searchfox.org/mozilla-central/source/toolkit/components/remote

**Reply by crocket** [reply]
For now, I have

```
ignore dbus-user
ignore private-tmp
```

in librewolf.local as a workaround.

**Reply by crocket** [reply]
I think `--join-or-start=librewolf` shouldn't create `librewolf-PID` sandboxes if multiple instances of librewolf are trying to start simultaneously, but if I wanted to work around the issue, I may create firejail-named-sandbox.

```
firejail-named-sandbox sandbox-name arguments-to-firejail...
```

This accepts commands on `$XDG_RUNTIME_DIR/firejail-named-sandbox` and sequentially processes firejail commands.

If there is no named sandbox, it creates one. If there is already the specified named sandbox, it joins the named sandbox.

**Reply by rusty-snake** [reply]
> I think --join-or-start=librewolf shouldn't create librewolf-PID sandboxes if multiple instances of librewolf are trying to start simultaneously

https://github.com/netblue30/firejail/issues/new

Maybe follow https://github.com/netblue30/firejail/blob/7b3cd31cfe77d5c4083ba38291567ea48effb1cb/.github/ISSUE_TEMPLATE/feature_request.md (not merged yet).

**Comment by rusty-snake** [comment]
I can not reproduce this using
```bash
i=0; while [[ $((i++)) -lt 100 ]]; do firefox --new-tab about:blank; done
```
but I can reproduce it using
```bash
firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank & firefox --new-tab about:blank
# or
cmd=(firefox --new-tab about:blank); ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]}
```

For me this looks like a deadlock, maybe caused by a race condition or broken/missing synchronization.

---

I can not reproduce this using

```console
$ bwrap --dev-bind / / --tmpfs $HOME --unshare-all /bin/bash
% /usr/bin/firefox >/dev/null 2>/dev/null & { sleep 20s; cmd=(/usr/bin/firefox --new-tab about:blank); ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]}; }
# or
$ firejail --noprofile --private
% /usr/bin/firefox >/dev/null 2>/dev/null & { sleep 20s; cmd=(/usr/bin/firefox --new-tab about:blank); ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]} & ${cmd[@]}; }
```

**Comment by crocket** [comment, marked as answer]
This discussion is moved to https://github.com/netblue30/firejail/issues/4543

**Comment by kmk3** [comment]
@crocket [on Sep 13](https://github.com/netblue30/firejail/discussions/4538#discussion-3568082):

> ```
> dbus-user.own org.mozilla.librewolf.*
> ```

Could you open a PR to add this to librewolf.profile?

**Reply by rusty-snake** [reply]
FWIW: I've it already in my local list with profile fixes.

**Reply by crocket** [reply]
@rusty-snake Will you submit the PR?

**Reply by crocket** [reply]
I submitted the PR.

**Reply by kmk3** [reply]
Link for reference: #4546
